amr-wind/physics/TerrainDrag.cpp

```diff
@@ -0,0 +1,110 @@
+#include "amr-wind/physics/TerrainDrag.H"
+#include "amr-wind/CFDSim.H"
+#include "AMReX_iMultiFab.H"
+#include "AMReX_MultiFabUtil.H"
+#include "AMReX_ParmParse.H"
+#include "AMReX_ParReduce.H"
+#include "amr-wind/utilities/trig_ops.H"
+#include "amr-wind/utilities/IOManager.H"
+
+namespace amr_wind::terraindrag {
+
+namespace {} // namespace
+
+TerrainDrag::TerrainDrag(CFDSim& sim)
+    : m_sim(sim)
+    , m_repo(sim.repo())
+    , m_mesh(sim.mesh())
+    , m_velocity(sim.repo().get_field("velocity"))
+    , m_terrainBlank(sim.repo().declare_field("terrainBlank", 1, 1, 1))
+    , m_terrainDrag(sim.repo().declare_field("terrainDrag", 1, 1, 1))
+{
+    std::ifstream file("terrain.amrwind");
+    amrex::Real value1, value2, value3;
+    while (file >> value1 >> value2 >> value3) {
+        m_xterrain.push_back(value1);
+        m_yterrain.push_back(value2);
+        m_zterrain.push_back(value3);
+    }
+    file.close();
+    m_sim.io_manager().register_io_var("terrainDrag");
+    m_sim.io_manager().register_io_var("terrainBlank");
+}
+
+void TerrainDrag::initialize_fields(int level, const amrex::Geometry& geom)
+{
+    using namespace utils;
+
+    const auto& dx = geom.CellSizeArray();
+    const auto& prob_lo = geom.ProbLoArray();
+    auto& velocity = m_velocity(level);
+    auto& blanking = m_terrainBlank(level);
+    auto& drag = m_terrainDrag(level);
+    // amrex::Print()<<" Terrain Initial:"<<level<<std::endl;
+    //  copy terrain data to gpu
+    amrex::Gpu::DeviceVector<amrex::Real> gpu_xterrain(m_xterrain.size());
+    amrex::Gpu::DeviceVector<amrex::Real> gpu_yterrain(m_xterrain.size());
+    amrex::Gpu::DeviceVector<amrex::Real> gpu_zterrain(m_xterrain.size());
+    auto* xterrain_ptr = gpu_xterrain.data();
+    auto* yterrain_ptr = gpu_yterrain.data();
+    auto* zterrain_ptr = gpu_zterrain.data();
+    amrex::ParallelFor(m_xterrain.size(), [=] AMREX_GPU_DEVICE(int n) {
```

Comment:
this won't work on device, you need a copy the way you did things in DragForcing.